src/nupic/algorithms/sdr_classifier.py

```diff
@@ -0,0 +1,375 @@
+      for (nSteps, votes) in retval.items():
+        if nSteps == "actualValues":
+          continue
+        print "    %d steps: " % (nSteps), _pFormatArray(votes)
+        bestBucketIdx = votes.argmax()
+        print ("      most likely bucket idx: "
+               "%d, value: %s" % (bestBucketIdx,
+                                  retval["actualValues"][bestBucketIdx]))
+      print
+
+    return retval
+
+
+
+  def infer(self, patternNZ, classification):
+    """
+    Return the inference value from one input sample. The actual
+    learning happens in compute(). The method customCompute() is here to
+    maintain backward compatibility.
+
+    Parameters:
+    --------------------------------------------------------------------
+    @param patternNZ  list of the active indices from the output below
+    @param classification dict of the classification information:
+                    bucketIdx: index of the encoder bucket
+                    actValue:  actual value going into the encoder
+
+    @return     dict containing inference results, one entry for each step in
+                self.steps. The key is the number of steps, the value is an
+                array containing the relative likelihood for each bucketIdx
+                starting from bucketIdx 0.
+
+                for example:
+                  {'actualValues': [0.0, 1.0, 2.0, 3.0]
+                    1 : [0.1, 0.3, 0.2, 0.7]
+                    4 : [0.2, 0.4, 0.3, 0.5]}
+    """
+
+    # Return value dict. For buckets which we don't have an actual value
+    # for yet, just plug in any valid actual value. It doesn't matter what
+    # we use because that bucket won't have non-zero likelihood anyways.
+
+    # NOTE: If doing 0-step prediction, we shouldn't use any knowledge
+    #  of the classification input during inference.
+    if self.steps[0] == 0:
+      defaultValue = 0
+    else:
+      defaultValue = classification["actValue"]
+    actValues = [x if x is not None else defaultValue
+                 for x in self._actualValues]
+    retval = {"actualValues": actValues}
+
+    for nSteps in self.steps:
+      predictDist = self.inferSingleStep(patternNZ, self._weightMatrix[nSteps])
+
+      retval[nSteps] = predictDist
+
+    return retval
+
+
+  def inferSingleStep(self, patternNZ, weightMatrix):
+    """
+    Perform inference for a single step. Given an SDR input and a weight
+    matrix, return a predicted distribution.
+
+    @param patternNZ  list of the active indices from the output below
+    @param weightMatrix numpy array of the weight matrix
+    @return numpy array of the predicted class label distribution
+    """
+    outputActivation = weightMatrix[patternNZ].sum(axis=0)
+
+    # softmax normalization
+    expOutputActivation = numpy.exp(outputActivation)
+    predictDist = expOutputActivation / numpy.sum(expOutputActivation)
+    return predictDist
+
+
+  def calculateError(self, classification):
```

Comment:
Is this part of the public interface? If it is only meant to be used internally then include a single underscore prefix: `_ calculateError`